executor/runcexecutor/executor.go

```diff
@@ -429,25 +430,122 @@ func (s *forwardIO) Stderr() io.ReadCloser {
 	return nil
 }
 
-// procHandle is to track the os process so we can send signals to it.
+// newProcKiller returns an abstraction for sending SIGKILL to the
+// process running inside the runc container.
+func newProcKiller(runC *runc.Runc, id string, isExec bool) (procKiller, error) {
+	if !isExec {
+		return procKiller{runC: runC, id: id}, nil
+	}
+
+	// for `runc exec` we need to create a pidfile and read it later to kill
+	// the process
+	tdir, err := os.MkdirTemp("", "runc")
+	if err != nil {
+		return procKiller{}, errors.Wrap(err, "failed to create directory for runc pidfile")
+	}
+
+	return procKiller{
+		runC:    runC,
+		id:      id,
+		pidfile: filepath.Join(tdir, "pidfile"),
+		cleanup: func() {
+			os.RemoveAll(tdir)
+		},
+	}, nil
+}
+
+type procKiller struct {
+	runC    *runc.Runc
+	id      string
+	pidfile string
+	cleanup func()
+}
+
+// Cleanup will delete any tmp files created for the pidfile allocation
+// if this killer was for a `runc exec` process.
+func (k procKiller) Cleanup() {
+	if k.cleanup != nil {
+		k.cleanup()
+	}
+}
+
+// Kill will send SIGKILL to the process running inside the container.
+// If the process was created by `runc run` then we will use `runc kill`,
+// otherwise for `runc exec` we will read the pid from a pidfile and then
+// send the signal directly that process.
+func (k procKiller) Kill(ctx context.Context) (err error) {
+	bklog.G(ctx).Debugf("sending sigkill to process in container %s", k.id)
+	defer func() {
+		if err != nil {
+			bklog.G(ctx).Errorf("failed to kill process in container id %s: %+v", k.id, err)
+		}
+	}()
+
+	if k.pidfile == "" {
+		// for `runc run` process we use `runc kill` to terminate the process
+		return k.runC.Kill(ctx, k.id, int(syscall.SIGKILL), nil)
+	}
+
+	// `runc exec` will write the pidfile a few milliseconds after we
+	// get the runc pid via the startedCh, so we might need to retry until
+	// it appears in the edge case where we want to kill a process
+	// immediately after it was created.
+	var pidData []byte
+	for {
+		pidData, err = os.ReadFile(k.pidfile)
+		if err != nil {
+			if os.IsNotExist(err) {
+				select {
+				case <-ctx.Done():
```

Comment:
I feel like it might be more robust to use a context like `waitCtx, cancel := context.WithTimeout(ctx, 10*time.Seconds)` or similar that's created in this function.

Reason being that we're now relying on the provided context to always have timeouts/cancellations setup properly, and even if that's true currently it seems like something that could be missed in any future modifications. So adding one extra layer of protection that's derived from the parent context probably doesn't hurt IMO.

This is a huge nitpick though, cases where this would matter are likely quite obscure.

Reply:
Maybe I am missing something, but I **think** this edge case is covered.  The ctx that is passed into the Kill function is created on like 564 like:
```
killCtx, timeout := context.WithTimeout(context.Background(), 7*time.Second)
```
This is the kill that is triggered on the request context being canceled.  There is also the client requested kill where client send in a SIGKILL on the Signal channel, that one will be using the cancelable Background context created in `runcProcessHandle` on line 543. So it is possible for the client requested SIGKILL via the Signal channel to block, but  the client will have a request context to cancel, which would trigger us to call kill again with the 7s timeout.

Reply:
Yeah I totally agree that at *this point* there wouldn't be a need for a timeout inside this function. I was just thinking that as changes to the rest of the code happen in the future it would be easy to miss something like this and accidentally provide a context that could potentially result in blocking.

But again, this is not important for functionality at this exact moment and a pretty obscure case to begin with, so no need to block merging on this point if you don't think it's worth it.

Reply:
I see what you are saying, seems fine to me, added it as a fail-safe.